test/e2e/storage/testsuites/snapshotsecrets.go

```diff
@@ -0,0 +1,246 @@
+			ginkgo.By("checking the PV")
+			_, err = cs.CoreV1().PersistentVolumes().Get(context.TODO(), pvc.Spec.VolumeName, metav1.GetOptions{})
+			framework.ExpectNoError(err)
+
+		}
+
+		cleanup := func() {
+			// Don't register an AfterEach then a cleanup step because the order
+			// of execution will do the AfterEach first then the cleanup step.
+			// Also AfterEach cleanup registration is not fine grained enough
+			// Adding to the cleanup steps allows you to register cleanup only when it is needed
+			// Ideally we could replace this with https://golang.org/pkg/testing/#T.Cleanup
+
+			// Depending on how far the test executed, cleanup accordingly
+			// Execute in reverse order, similar to defer stack
+			for i := len(cleanupSteps) - 1; i >= 0; i-- {
+				err := tryFunc(cleanupSteps[i])
+				framework.ExpectNoError(err, "while running cleanup steps")
+			}
+
+		}
+		ginkgo.BeforeEach(func() {
+			init()
+		})
+		ginkgo.AfterEach(func() {
+			cleanup()
+		})
+
+		ginkgo.Context("", func() {
+			var (
+				vs     *unstructured.Unstructured
+				secret *v1.Secret
+			)
+
+			ginkgo.BeforeEach(func() {
+				var sr *SnapshotResource
+				cleanupSteps = append(cleanupSteps, func() {
+					framework.ExpectNoError(sr.CleanupResource())
+				})
+				// create a secret
+				secret = &v1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: f.Namespace.Name,
+						Name:      "snapshot-secret",
+					},
+					Data: map[string][]byte{
+						"secret-data": []byte("secret-value-1"),
+					},
+				}
+
+				if secret, err = cs.CoreV1().Secrets(f.Namespace.Name).Create(context.TODO(), secret, metav1.CreateOptions{}); err != nil {
+					framework.Failf("unable to create test secret %s: %v", secret.Name, err)
+				}
+
+				// Create snapshot using secret
+				parameters := map[string]string{
+					CSISnapshotterSecretName:      "snapshot-secret",
+					CSISnapshotterSecretNamespace: f.Namespace.Name,
+				}
+				sr = CreateSnapshotResource(sDriver, config, pattern, pvc.GetName(), pvc.GetNamespace(), parameters)
+				vs = sr.Vs
+			})
+			ginkgo.It("Use secrets field, create/list and delete snapshots", func() {
+
+				ginkgo.By("list the snapshot")
+				// TODO: Check if list snapshot is supported. Log and continue if not supported.
+				ListSnapshotsWithSecrets(dc, vs.GetNamespace(), vs.GetName(), framework.Poll, framework.SnapshotDeleteTimeout)
+
+				ginkgo.By("Delete the snapshot")
+				DeleteSnapshotWithSecrets(cs, dc, vs.GetNamespace(), vs.GetName(), framework.Poll, framework.SnapshotDeleteTimeout)
+				ginkgo.By("checking if the snapshot is deleted")
+
+			})
+		})
+	})
+}
+
+// ListSnapshotsWithSecrets lists a given volumesnapshot it uses secret provided within storageclass
+func ListSnapshotsWithSecrets(dc dynamic.Interface, ns string, snapshotName string, poll, timeout time.Duration) error {
+	ginkgo.By("listing the snapshot")
+	_, err := dc.Resource(SnapshotGVR).Namespace(ns).List(context.TODO(), metav1.ListOptions{})
```

Comment:
@xing-yang can you please point me to the proper options to be passed for `ListOptions` for secrets? Couldn't find it documented anywhere.

Reply:
ListSnapshots is only called by the CSI snapshotter sidecar for pre-provisioned snapshots.  It is optional so snapshot function will not fail whether it is supported or not.  It is not envoked by the K8S API List call.
See this PR: https://github.com/kubernetes-csi/external-snapshotter/pull/252

I don't think you can test much in e2e test, but we have lots of unit tests in external-snapshotter already.  In this e2e test, just add secrets for ListSnapshots and test it as part of the general test for pre-provisioned snapshots.

Reply:
>  just add secrets for ListSnapshots

Is there a way to do this using k8s API call?

Reply:
ListSnapshots CSI RPC is triggered when you create a VolumeSnapshot API object using k8s API.